Userland/Libraries/LibWeb/HTML/HTMLOptionsCollection.cpp

```diff
@@ -33,6 +35,41 @@ void HTMLOptionsCollection::initialize(JS::Realm& realm)
     WEB_SET_PROTOTYPE_FOR_INTERFACE(HTMLOptionsCollection);
 }
 
+// https://html.spec.whatwg.org/multipage/common-dom-interfaces.html#dom-htmloptionscollection-length
+WebIDL::ExceptionOr<void> HTMLOptionsCollection::set_length(WebIDL::UnsignedLong value)
+{
+    // 1. Let current be the number of nodes represented by the collection.
+    auto current = length();
+
+    // 2. If the given value is greater than current, then:
+    if (value > current) {
+        // 2.1. If the given value is greater than 100,000, then return.
+        if (value > 100'000)
+            return {};
+
+        // 2.2. Let n be value − current.
+        auto n = value - current;
+
+        // 2.3. Append n new option elements with no attributes and no child nodes to the select element on which this is rooted.
+        // Mutation events must be fired as if a DocumentFragment containing the new option elements had been inserted.
+        auto root_element = root();
+        for (size_t i = 0; i < n; i++)
+            TRY(root_element->append_child(TRY(DOM::create_element(root_element->document(), HTML::TagNames::option, Namespace::HTML))));
+    }
+
+    // 3. If the given value is less than current, then:
+    if (value < current) {
+        // 3.1. Let n be current − value.
+        auto n = current - value;
+
+        // 3.2. Remove the last n nodes in the collection from their parent nodes.
+        for (size_t i = current - n; i < current; i++)
+            m_cached_elements[i]->remove();
```

Comment:
This use of `m_cached_nodes` looks very footgunny / prone to breakage. `m_cached_elements` is being marked as `protected`,  but `update_cache_if_needed` is not. Currently this works because we're calling length, and that will update the cache - but this does feel too brittle.

I'm not sure 100% sure if this cache should be exposed publically, or if we should find a convient wrapper function / callback for this.

Reply:
Probably we could also use `item()` in this case too

Reply:
True it was not needed I will use `item()`